lib/http_aws_sigv4.c

```diff
@@ -924,4 +986,205 @@ CURLcode Curl_output_aws_sigv4(struct Curl_easy *data)
+        }
+        segment_length = 0;
+        index++;
+        num_splits++;
+      }
+      start = pos + 1;
+    }
+    else {
+      segment_length++;
+    }
+  }
+
+  if(segment_length) {
+    if(num_splits == MAX_QUERY_COMPONENTS) {
+      goto fail;
+    }
+    Curl_dyn_init(&db[index], segment_length + 1);
+    result = Curl_dyn_addn(&db[index], &source[start],
+      segment_length);
+    if(result) {
+      goto fail;
+    }
+    num_splits++;
+  }
+fail:
+  *num_splits_out = num_splits;
+  return result;
+}
+
+
+static bool is_reserved_char(const char c)
+{
+  return (ISALNUM(c) || ISURLPUNTCS(c));
+}
+
+static CURLcode uri_encode_path(struct Curl_str *original_path,
+struct dynbuf *new_path) {
+
+  const char *p = Curl_str(original_path);
+  CURLcode result = CURLE_OK;
+  size_t index;
+
+  for(index = 0; index < Curl_strlen(original_path); index++) {
+    /* Do not encode slashes or unreserved chars from RFC 3986 */
+    unsigned char c = p[index];
+    if(is_reserved_char(c) || c == '/') {
+      result = Curl_dyn_addn(new_path, &c, 1);
+      if(result) {
+        goto fail;
+      }
+    }
+    else {
+      result = Curl_dyn_addf(new_path, "%%%02X", c);
+      if(result) {
+        goto fail;
+      }
+    }
+  }
+fail:
+  return result;
+}
+
+
+static CURLcode encode_query_component(char *component, size_t len,
+  struct dynbuf *db) {
+
+  size_t index;
+  CURLcode result = CURLE_OK;
+  unsigned char this_char;
+
+  for(index = 0; index < len; index++) {
+
+    this_char = component[index];
+
+    if(is_reserved_char(this_char)) {
+      /* Escape unreserved chars from RFC 3986 */
+      result = Curl_dyn_addn(db, &this_char, 1);
+      if(result) {
+        goto fail;
+      }
+    }
+    else if(this_char == '+') {
+      /* Encode '+' as space */
+      result = Curl_dyn_add(db, "%20");
+      if(result) {
+        goto fail;
+      }
+    }
+    else {
+      result = Curl_dyn_addf(db, "%%%02X", this_char);
+      if(result) {
+        goto fail;
+      }
+    }
+  }
+fail:
+  return result;
+}
+
+/*
+* Populates a dynbuf containing url_encode(url_decode(in))
+*/
+
+static CURLcode http_aws_decode_encode(const char *in, size_t in_len,
+struct dynbuf *out) {
+  CURLcode result = CURLE_OK;
+  char *out_s;
+  size_t out_s_len;
+
+  result = Curl_urldecode(in, in_len,
+    &out_s, &out_s_len,
+    REJECT_NADA);
+  if(result) {
+    goto fail;
+  }
+  result = encode_query_component(out_s, out_s_len, out);
+  Curl_safefree(out_s);
+fail:
+  return result;
+}
+
+static bool should_urlencode(struct Curl_str *service_name)
+{
+  /*
+   * These services require unmodified (not additionally url encoded) URL
+   * paths.
+   * should_urlencode == true is equivalent to should_urlencode_uri_path
+   * from the AWS SDK. Urls are already normalized by the curl url parser
+   */
+
+  if(!strcmp(Curl_str(service_name), "s3") ||
+    !strcmp(Curl_str(service_name), "s3-express") ||
+    !strcmp(Curl_str(service_name), "s3-outposts")) {
```

Comment:
Should these perhaps rather use `Curl_str_cmp()` ? I presume they need a full match and not just a prefix?